Prepare for conan package / Add ci jobs for ubuntu, mac and windows

phbasler opened the pull request:
In preparation for the conan package I added the options to get the necessary dependencies through conan. 
You can build cmaes with conan, as shown in the workflow.

While configuring this I saw that the PythonInterp find_package call was deprecated and so I updated it to find_package(Python3) with the modules. This also allowed me to make the python and t_* tests run with ctest. 
You can now just call ctest in your build directory and the 6 tests run. 

Furthermore, we can get rid of the custom FindNumpy and FindEigen module. cmake has it's own find modules for eigen and python.

phbasler commented:
I just saw that it still does not compile on windows with mvsc. I convert this PR to a draft until I fixed it as well.

phbasler commented:
@beniz  It would not compile on windows until I defined an explicit copy ctr for scalar_normal_dist_op. Can you check if this one is ok? 

ATM the python tests are failing because the module is not found for the ctest call, but everything compiles.

phbasler commented:
@beniz @nikohansen 

I found the problem and maybe one of you has an idea. 
On my fork https://github.com/phbasler/libcmaes/tree/conan-windows I set up some CI to run the unit tests, including the python tests, on windows, mac and ubuntu. 
Mac and ubuntu work without a problem, but on windows the lcames python module is seperated in two files. One build/python/Release/lcames.pyd and the dll build/src/Release/cmaes.dll. It works there if I copy the cmaes.dll to the directory which contains the .pyd. If I don't copy it, then I get an Import Error because the dll is not found. 

What should we do here? Do you have an idea on how to fix it or should we copy it during the cmake call?

nikohansen commented:
Sorry, I have no idea regarding the windows test setup.

beniz commented:
Hi @phbasler I would go for the copy of the DLL is that helps. Thanks for all the work on this!

phbasler commented:
@beniz  That's how I implemented now. I also fixed a small mistake I introduced with another PR:
If you include libcmaes as a subdirectory then the doc target was also created, even if you have one of your one (like I have in https://github.com/phbasler/BSMPT) and then I got a cmake error there. I fixed that with the last commit that the doc target is only created if libcmaes is build as the top project

phbasler commented:
@nikohansen @beniz I finished the conan package. You can build it with `conan create conanfile.py --profile=SomeProfile" where SomeProfile is a conan profile with your compiler and settings how you want to build. You need to fill in the author and topics in the conanfile.

beniz commented:
@phbasler hello, apologies for missing this, you should put your name in the conan file! 
I am not familiar with conan, if it works for you, then maybe squash and get ready for a merge!

phbasler commented:
Seems that there are some changes necessary. I will have a look at it if I have some time to spare, but this may take a while.

phbasler commented:
@beniz @nikohansen 
So after some time I got around to update this PR and now it works with conan 2 
If we can go through this PR and you can release a new version, then I can make the official conan recipe

phbasler commented:
@beniz just a reminder that this is still open

phbasler commented:
@beniz I removed the windows unit tests. There is some extra work required with the boost python setup so the tests fails. But the normal conan create build runs including the test_package, so we see that it still works on windows

phbasler commented:
@beniz I think this is blocking the conan PR.
Any chance we can finish this before the PR reaches the 3 year mark?

phbasler commented:
@beniz  @nikohansen 
I saw that other people basically did the same work with #246  and #247 again, maybe we can finish this at some point? I opened this PR now almost 3 years ago

beniz commented:
Hi @phbasler sorry I have no windows experience or knowhow. If you recommend something that works and can be merged, I'll follow your lead.

phbasler commented:
@beniz Please have a look at the CI jobs.
Everything is being build already in the CI on windows.

The only thing not working in windows is the unit tests. This is because the boost python setup still needs adjustments on Windows.
The normal library can be build and used without problems on Windows.

phbasler commented:
@beniz Any news here?

phbasler commented:
@beniz Everything is ready for review. If you can review it, that would be great
